polyfill/intersectionobserver-polyfill.js

```diff
@@ -0,0 +1,250 @@
+          boundingClientRect: targetRect,
+          intersectionRect: intersectionRect,
+          intersectionRatio: intersectionRatio,
+          target: target
+        };
+        Object.freeze(intersectionEntry);
+        this._queue.push(intersectionEntry);
+        this._scheduleCallback();
+        this._observationTargets.set(target, intersectionEntry);
+      }.bind(this));
+    },
+
+    _scheduleCallback: function() {
+      if(this._timeoutId) {
+        return;
+      }
+      this._timeoutId = scope.setTimeout(function() {
+        this._descheduleCallback();
+        this._callback(this._queue, this);
+        this._queue = [];
+      }.bind(this), POLL_INTERVAL);
+    },
+
+    _descheduleCallback: function() {
+      if(!this._timeoutId) {
+        return;
+      }
+      scope.clearTimeout(this._timeoutId);
+      this._timeoutId = null;
+    },
+
+    _rootRect: function() {
+      if(this._root) {
+        return getBoundingClientRect(this.root);
+      }
+      return {
+        top: 0,
+        left: 0,
+        right: scope.innerWidth,
+        width: scope.innerWidth,
+        bottom: scope.innerHeight,
+        height: scope.innerHeight
+      };
+    },
+
+    _hasCrossedThreshold: function(oldRatio, newRatio) {
+      if(oldRatio === newRatio) {
+        return;
+      }
+      var isOnBoundary = this.thresholds.map(function(threshold) {
+        return threshold == newRatio;
+      }).indexOf(true) !== -1;
+
+      var b1 = this.thresholds.map(function(threshold) {
+        return threshold < oldRatio;
+      });
+      var b2 = this.thresholds.map(function(threshold) {
+        return threshold < newRatio;
+      });
+      var hasCrossedThreshold = b1.map(function(_, idx) {
+        return b1[idx] !== b2[idx]
+      }).indexOf(true) !== -1;
+      return isOnBoundary || hasCrossedThreshold;
+    }
+  };
+
+  var intersectRects = function(rect1, rect2) {
+    var top = Math.max(rect1.top, rect2.top);
+    var bottom = Math.min(rect1.bottom, rect2.bottom);
+    var left = Math.max(rect1.left, rect2.left);
+    var right = Math.min(rect1.right, rect2.right);
+    var intersectionRect = {
+      top: top,
+      bottom: bottom,
+      left: left,
+      right: right,
+      width: right - left,
+      height: bottom - top
+    };
+    if(top > bottom) {
+      intersectionRect.height = 0;
+    }
+    if(left > right) {
+      intersectionRect.width = 0;
+    }
+    return intersectionRect;
+  };
+
+  scope.IntersectionObserver = IntersectionObserver;
+
+  var throttle = function(fn, int) {
+    var timer = null;
+    return function () {
+      if (timer) {
+        return;
+      }
+      timer = setTimeout(function () {
+        fn.apply(this, arguments);
+        timer = null;
+      }.bind(this), int);
+    };
+  };
+
+  var getBoundingClientRect = function(el) {
+    var r = el.getBoundingClientRect();
+    if(!r) {
+      return null;
+    }
+    // Older IE
+    r.width = r.width || r.right - r.left;
```

Comment:
Safari throws an read-only error on setting this property. I fixed it by checking if the property is writable using the property descriptor getter.

```javascript
var desc = Object.getOwnPropertyDescriptor(r, 'width');
if (desc && desc.writable) {
  r.width = r.width || r.right - r.left;
  r.height = r.height || r.bottom - r.top;
}
```